homeassistant/components/hlk_sw16.py

```diff
@@ -0,0 +1,200 @@
+        switches = config[DOMAIN][device][CONF_SWITCHES]
+
+        host = config[DOMAIN][device][CONF_HOST]
+        port = config[DOMAIN][device][CONF_PORT]
+
+        @callback
+        def disconnected():
+            """Schedule reconnect after connection has been lost."""
+            _LOGGER.warning('HLK-SW16 disconnected')
+            async_dispatcher_send(hass, SIGNAL_AVAILABILITY, False)
+
+        @callback
+        def reconnected():
+            """Schedule reconnect after connection has been lost."""
+            _LOGGER.warning('HLK-SW16 connected')
+            async_dispatcher_send(hass, SIGNAL_AVAILABILITY, True)
+
+        async def connect():
+            """Set up connection and hook it into HA for reconnect/shutdown."""
+            _LOGGER.info('Initiating HLK-SW16 connection')
+
+            client = await create_hlk_sw16_connection(
+                host=host,
+                port=port,
+                disconnect_callback=disconnected,
+                reconnect_callback=reconnected,
+                loop=hass.loop,
+                logger=_LOGGER,
+                timeout=CONNECTION_TIMEOUT,
+                reconnect_interval=DEFAULT_RECONNECT_INTERVAL)
+
+            hass.data[DATA_DEVICE_REGISTER][device] = client
+
+            # Load platforms
+            for comp_name, comp_conf in switches.items():
+                hass.async_create_task(
+                    async_load_platform(hass, 'switch', DOMAIN,
+                                        (comp_name, comp_conf, device),
+                                        config))
+
+            # handle shutdown of HLK-SW16 asyncio transport
+            hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP,
+                                       lambda x: client.stop())
+
+            _LOGGER.info('Connected to HLK-SW16')
+
+        hass.loop.create_task(connect())
+
+    for device in config[DOMAIN]:
+        add_device(device)
+    return True
+
+
+class SW16Device(Entity):
+    """Representation of a HLK-SW16 device.
+
+    Contains the common logic for HLK-SW16 entities.
+    """
+
+    def __init__(self, relay_name, device_port, device_id, client):
+        """Initialize the device."""
+        # HLK-SW16 specific attributes for every component type
+        self._device_id = device_id
+        self._device_port = device_port
+        self._is_on = None
+        self._client = client
+        self._name = relay_name
+
+    @callback
+    def handle_event_callback(self, event):
+        """Propagate changes through ha."""
+        _LOGGER.debug("Relay %s new state callback: %r",
+                      self._device_port, event)
+        self._is_on = event
+        self.async_schedule_update_ha_state()
+
+    @property
+    def should_poll(self):
+        """No polling needed."""
+        return False
+
+    @property
+    def name(self):
+        """Return a name for the device."""
+        return self._name
+
+    @property
+    def is_on(self):
```

Comment:
Only binary sensors or actuators implement `is_on`. If this class should be common we should only implement common entity properties, or integration specific things. Since we only have switches as of yet, maybe just move this class to the switch platform and merge it with that class?

Alternatively move this property to the switch entity class.

Reply:
Moved to `SW16Switch`